### ISSUE TITLE
Incompatibility with django-channels-graphql-ws

### ISSUE THREAD

**Issue by rakyi** [CJWorkbench/channels_rabbitmq#3]
When using channels-rabbitmq together with django-channels-graphql-ws I get the following errors when trying to initialize a GraphQL subscription:

```
An error occurred while resolving field GraphQLSubscription.notification
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/graphql/execution/executor.py", line 447, in resolve_or_error
    return executor.execute(resolve_fn, source, info, **args)
  File "/usr/local/lib/python3.7/site-packages/graphql/execution/executors/asyncio.py", line 69, in execute
    result = fn(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py", line 466, in _subscribe
    return register(groups, publish_callback, unsubscribed_callback)
  File "/usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py", line 889, in register
    group, self.channel_name
  File "/usr/local/lib/python3.7/site-packages/asgiref/sync.py", line 64, in __call__
    return call_result.result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.7/site-packages/asgiref/sync.py", line 78, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/channels_rabbitmq/core.py", line 183, in group_add
    await connection.group_add(group, channel)
  File "/usr/local/lib/python3.7/site-packages/channels_rabbitmq/connection.py", line 268, in inner
    return await fn(self, self._channel, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/channels_rabbitmq/connection.py", line 628, in group_add
    ), "This layer won't let you add a channel from another connection"
AssertionError: This layer won't let you add a channel from another connection
Task exception was never retrieved
future: <Task finished coro=<GraphqlWsConsumer.receive_json.<locals>.process_message() done, defined at /usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py:699> exception=KeyError('GRAPHQL_WS_SUBSCRIPTION-91a52ed21b5a46b42f1b42b8360df62d')>
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py", line 716, in process_message
    await self._on_gql_stop(operation_id=content["id"])
  File "/usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py", line 1013, in _on_gql_stop
    assert self._sids_by_group[group].count(operation_id) == 1, (
KeyError: 'GRAPHQL_WS_SUBSCRIPTION-91a52ed21b5a46b42f1b42b8360df62d'
```

The same thing works with channels-redis. I'm wondering what's the reason behind this limitation: https://github.com/CJWorkbench/channels_rabbitmq/blob/da15d0cfcd8b056a04e8c1958c04a828f88aa9cf/channels_rabbitmq/connection.py#L626-L628

**Comment by adamhooper** [comment]
I've looked through both projects' code bases and I think I understand the problem.

TL;DR this won't be fixed in channels-rabbitmq.

*Why the limitation*

Unlike channels-redis, channels-rabbitmq doesn't let one RabbitMQ connection create a subscription on behalf of a _different_ RabbitMQ connection. That's because RabbitMQ was made for this sort of thing, and Redis wasn't.

Redis is just a blob of shared memory. A channels-redis "subscription" is a shared list; it isn't associated with any connection. When the connection goes away, the list stays in memory. Any connection can "subscribe" to any other connection's list. Simple, right? Not quite. In production, there are big problems: lists leak when connections drop, and races abound. (I created channels_rabbitmq because channels_redis was stalling.)

RabbitMQ is actually designed for passing messages. A subscription is tied to a connection. When the connection closes or dies, RabbitMQ (the _server_) cleans up all resources associated with that connection. One client doesn't get to subscribe a _different_ client to messages, because that just isn't a feature anybody wants. It's like a comic-book villain of features: transparently evil and the world would obviously be better without it.

Rest assured: there are real underlying bugs in django-channels-graphql-ws that this limitation prevents. They probably involve unwanted messages, lost messages, unwanted subscriptions, lost subscriptions, leaked Redis memory and/or stalls. It would waste everybody's time to debug them.

*How to fix it*

django-channels-graphql-ws should open a single connection to RabbitMQ, instead of one connection per worker thread. (Really, it should use the one event loop, instead of one event loop per worker.) I know that's hard, but concurrent programming is hard!

**Comment by adamhooper** [comment]
Closing as wontfix because it's impossible to change this behavior with RabbitMQ.
